### PR TITLE
Cloud building, configuration and build request management

### PR DESCRIPTION
**Pull request by blckmn**
This is a preliminary cut, and the cloud build options have yet to be implemented in the UI.

Open to ideas on how we should implement the design of asking for various radio and telemetry protocols.

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by sugaarK** [comment]
@limonspb @haslinghuis @McGiverGim @Asizon

File: src/tabs/firmware_flasher.html
Comment (haslinghuis) [inline src/tabs/firmware_flasher.html:96]: Please move closing div to next line.

File: src/js/tabs/firmware_flasher.js
Comment (haslinghuis) [reply src/js/tabs/firmware_flasher.js:4]: Other suggestions (fixes autodetect):

```
line 88:            if (self.targets) {
line 489:           const isLoaded = self.targets ? Object.keys(self.targets).length > 0 : false;
```

File: src/js/tabs/firmware_flasher.js
Comment (haslinghuis) [inline src/js/tabs/firmware_flasher.js:141]:
```suggestion
                .sort((a,b) => a.target - b.target)
```

Sort works on the first part of the list now, but at the end there are still some legacy targets unsorted.

**Review by limonspb — COMMENTED** [review]
Why there is cost self=this everywhere? Can just use this directly, can't we?

File: src/main.html
Comment (limonspb) [inline src/main.html:113]: @chmelevskij will be mad that new module does not use modules :)

Comment (blckmn) [reply src/main.html:113]: I'm not yet familiar... happy to revert in due course. This was the fastest way home for me.

**Comment by haslinghuis** [comment]
As we already using `select2` we should be looking at something like:

image

https://select2.org/appearance

File: src/main.html
Comment (haslinghuis) [reply src/main.html:113]: It's ok for now. No need to revert.

**Comment by haslinghuis** [comment]
Wow!!!

File: src/js/release_loader.js
Comment (chmelevskij) [inline src/js/release_loader.js:21]: Why not using `class` here? `prototype` hacking is sort of thing of the past for new code.

File: src/js/tabs/firmware_flasher.js
Comment (chmelevskij) [inline src/js/tabs/firmware_flasher.js:5]: Wondering what is the risk here of dropping old way to get firmware? Should we keep both at least for first release/experimental flag?

File: src/js/ConfigInserter.js
Comment (chmelevskij) [inline src/js/ConfigInserter.js:23]: Where these only used for old firmware flashing?

**Review by chmelevskij — COMMENTED** [review]
Just some questions about removing the old one straight away

**Comment by blckmn** [comment]
> Just some questions about removing the old one straight away

I've copied the old hexes to the new build API. It supports downloading of anything after 4.0.0.

File: src/js/tabs/firmware_flasher.js
Comment (blckmn) [reply src/js/tabs/firmware_flasher.js:5]: The new system supports the old hexs.

File: src/js/ConfigInserter.js
Comment (blckmn) [reply src/js/ConfigInserter.js:23]: The hexes get updated by the build API before downloading.

**Comment by limonspb** [comment]
A few questions:
1. Is it planned for 4.4?
2. Is there a way to flash Dev build in the old manner?
3. Is there a way to flash different dev builds like it was before?
image

**Comment by blckmn** [comment]
400 means the build failed, and you can check the log.

It will support passing any commit hash. Just need to think of the UI for that. You can even pass a pull request reference...

**Comment by sugaarK** [comment]
@haslinghuis  @limonspb  @KarateBrot  need this merged

**Review by haslinghuis — APPROVED** [review]
Another class 😃

**Comment by haslinghuis** [comment]
How about the TODO in line 801 (firmware_flasher.js)

File: src/js/release_loader.js
Comment (chmelevskij) [inline src/js/release_loader.js:3]: ❤️

**Comment by haslinghuis** [comment]
image

**Comment by blckmn** [comment]
> How about the TODO in line 801 (firmware_flasher.js)

Left it in there as it's something to think about. It was as it was before but I question it's need... for a future review and cleanup.

**Comment by haslinghuis** [comment]
https://build.betaflight.com/api/builds/4.4.0/MATEKH743/48b7b958-fe78-a5a5-ffb1-a91193fd5bd4/log

Flashing fails on verifying:
```
Erase: complete
stm32usbdfu.js:924 Writing data ...
stm32usbdfu.js:982 Writing: done
stm32usbdfu.js:996 Verifying data ...
common.js:39 Chrome API Error: Transfer failed..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:39:90)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:451:17
checkChromeRuntimeError @ common.js:39
(anonymous) @ stm32usbdfu.js:451
stm32usbdfu.js:452 USB controlTransfer IN failed for request 2!
stm32usbdfu.js:454 USB transfer result code: 1
common.js:39 Chrome API Error: Transfer failed..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:39:90)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:451:17
checkChromeRuntimeError @ common.js:39
(anonymous) @ stm32usbdfu.js:451
stm32usbdfu.js:452 USB controlTransfer IN failed for request 2!
stm32usbdfu.js:454 USB transfer result code: 1
common.js:39 Chrome API Error: No such connection..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:39:90)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:451:17
checkChromeRuntimeError @ common.js:39
(anonymous) @ stm32usbdfu.js:451
stm32usbdfu.js:452 USB controlTransfer IN failed for request 2!
main.html#:1 Error handling response: TypeError: Cannot read properties of undefined (reading 'resultCode')
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:454:24
port_handler.js:276 PortHandler - Found: [{"path":"/dev/ttyACM0","displayName":"Betaflight STM32H743","vendorId":1155,"productId":22336}]
```

**Comment by haslinghuis** [comment]
https://build.betaflight.com/api/builds/4.4.0/MATEKH743/48b7b958-fe78-a5a5-ffb1-a91193fd5bd4/log

```
Erase: complete
stm32usbdfu.js:924 Writing data ...
stm32usbdfu.js:982 Writing: done
stm32usbdfu.js:996 Verifying data ...
common.js:39 Chrome API Error: Transfer failed..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:39:90)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:451:17
checkChromeRuntimeError @ common.js:39
(anonymous) @ stm32usbdfu.js:451
stm32usbdfu.js:452 USB controlTransfer IN failed for request 2!
stm32usbdfu.js:454 USB transfer result code: 1
common.js:39 Chrome API Error: Transfer failed..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:39:90)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:451:17
checkChromeRuntimeError @ common.js:39
(anonymous) @ stm32usbdfu.js:451
stm32usbdfu.js:452 USB controlTransfer IN failed for request 2!
stm32usbdfu.js:454 USB transfer result code: 1
common.js:39 Chrome API Error: No such connection..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:39:90)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:451:17
checkChromeRuntimeError @ common.js:39
(anonymous) @ stm32usbdfu.js:451
stm32usbdfu.js:452 USB controlTransfer IN failed for request 2!
main.html#:1 Error handling response: TypeError: Cannot read properties of undefined (reading 'resultCode')
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:454:24
port_handler.js:276 PortHandler - Found: [{"path":"/dev/ttyACM0","displayName":"Betaflight STM32H743","vendorId":1155,"productId":22336}]
```

**Comment by haslinghuis** [comment]
If there is no internet connection we get:

image

File: src/js/release_loader.js
Comment (haslinghuis) [inline src/js/release_loader.js:40]:
```suggestion
                if (onFailure !== undefined) {
```

File: src/js/release_loader.js
Comment (haslinghuis) [inline src/js/release_loader.js:77]:
```suggestion
            if (typeof(onFailure) !== undefined) {
```

File: src/js/release_loader.js
Comment (haslinghuis) [inline src/js/release_loader.js:98]:
```suggestion
            if (typeof(onFailure) !== undefined) {
```

File: src/js/release_loader.js
Comment (haslinghuis) [inline src/js/release_loader.js:112]:
```suggestion
            if (typeof(onFailure) !== undefined) {
```

File: src/js/tabs/firmware_flasher.js
Comment (haslinghuis) [inline src/js/tabs/firmware_flasher.js:134]:
```suggestion
            if (!targets || !navigator.connection) {
```

**Review by haslinghuis — CHANGES_REQUESTED** [review]
Please to fix some errors when there is no internet connection add:

image

File: src/css/tabs/firmware_flasher.less
Comment (haslinghuis) [inline src/css/tabs/firmware_flasher.less:128]: Remove the 30px gap:

```suggestion
            margin-left: auto;
            color: #3f4241;
```

File: src/js/tabs/firmware_flasher.js
Comment (haslinghuis) [inline src/js/tabs/firmware_flasher.js:176]: Sorry almost forgot this one:

```suggestion
        function buildOptions(data) {
            if (navigator.connection) {
                buildOptionsList($('select[name="radioProtocols"]'), data.radioProtocols);
                buildOptionsList($('select[name="telemetryProtocols"]'), data.telemetryProtocols);
                buildOptionsList($('select[name="options"]'), data.generalOptions);
            }
```

**Comment by blckmn** [comment]
> Please to fix some errors when there is no internet connection add:
> 

@haslinghuis changes as requested.

File: src/js/tabs/firmware_flasher.js
Comment (blckmn) [reply src/js/tabs/firmware_flasher.js]: Sorted.

**Review by haslinghuis — APPROVED** [review]

**Review by haslinghuis — APPROVED** [review]

**Review by limonspb — CHANGES_REQUESTED** [review]
I think this is an awesome piece of work for sure and we should continue improving/supporting it.
Love to see that we don't show release notes anymore in Configurator.
Guys, but i have to be defending users here :)
While i like the idea of cloud building, we have to take into account a few things.

Feature freeze is coming in a few days. If we merge this PR, do we still have time to improve it? I guess we can't during the feature freeze unless its a bugfix. Can't bring more features to it.

So im looking at this PR as an official 10.9 release and thinking about the general user.

Let's not make the flashing process more complicated than it is now for the general user.
I understand that some users will HAVE to deal with the Build Configuration because of the F4/F7 flash limitations if they want some special settings to be included/excluded.

However, the general user-noob should be able to just select 4.4, autodetect target and click flash like he was able to do so before, without selecting from a bunch of "weird abbreviations".
While he does that, on the backend the cloud building could be happening with some presets - ok. But we can't show/overwhelm the general "noob" with all these extra options unless he wants to mess with them.

I understand that in future we are planning to add a default preset that will sort of solve the situation. BUT:
1. This preset should be sort of invisible for the general user that just flashes and wants to fly regular drone.
2. **Currently impossible to flash different development versions from online like before.**

So we have a few options:
1. Schedule cloud build for 4.5, not 4.4, then we have time to address the current concerns.
2. Address these concerns in this PR
3. Take a word from @blckmn that he will address these concerns in the next couple of days in the separate PR that we will merge for 4.4 :)
4. Move the feature freeze date so we have time for finalizing this.

**Comment by sugaarK** [comment]
@limonspb  Wana take this to the management channel on slack

**Comment by limonspb** [comment]
> @limonspb Wana take this to the management channel on slack

yeah sure, whatever works
I just saw that @haslinghuis approved the PR. Wanted to make sure it won't get fast-merged without you guys reading my complains :)

**Comment by sugaarK** [comment]
You have to remember @limonspb while you have valid points I’m pretty sure not very one agrees with not merging this and getting it dialled in  the rc phase is perfect for testing this and we already made the decision that this aspect of the project wouldn’t be constrained by the feature freeze

**Comment by limonspb** [comment]
> You have to remember @limonspb while you have valid points I’m pretty sure not every one agrees with not merging this and getting it dialled in during the rc phase is perfect for testing this and we already made the decision that this aspect of the project wouldn’t be constrained by the feature freeze

So are we allowed to add more features/UI changes to the Cloud Building during the feature freeze? That's a little weird :)
I honestly don't see a rush for this feature for right now. All targets fitting all the features, just like in 4.3. it's great for future for sure. But no need to compromise the usability just cuz we want to rush with could builds

**Comment by sugaarK** [comment]
There will be a dead simple version of the flashers for the noobs that you pretty much only choose your radio protocol. As it is thanks to the defines in the targets now, your already winning on the space side of things using the new builder as it only includes the hardware libraries that are defined in the target. .. 

Start a thread in the slack so we can work it out with the guys.. use the works channel every one can weigh in on it

**Comment by limonspb** [comment]
> There will be a dead simple version of the flashers for the noobs that you pretty much only choose your radio protocol. As it is thanks to the defines in the targets now, your already winning on the space side of things using the new builder as it only includes the hardware libraries that are defined in the target. ..
> 
> Start a thread in the slack so we can work it out with the guys.. use the works channel every one can weigh in on it

Yeah it already started :)

**Review by sugaarK — APPROVED** [review]

**Review by haslinghuis — DISMISSED** [review]

**Comment by haslinghuis** [comment]
@blckmn should we add semver for 1.45 adding cloud build to release and release candidates too?

**Review by sugaarK — APPROVED** [review]

**Review by haslinghuis — APPROVED** [review]

**Review by sugaarK — APPROVED** [review]

**Review by limonspb — APPROVED** [review]

**Pull request merged**
